File: LC/Medium/739.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        n = len(temperatures)
        ans = [0]*n
        stack = []

        for curr_day in range(n):
            while stack and temperatures[stack[-1]] < temperatures[curr_day]:
                past_day = stack.pop()
                ans[past_day] = curr_day - past_day

            stack.append(curr_day)

        return ans

File: LC/C++/Medium/739.cpp
#include <array>
#include <vector>

class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        const int n = temperatures.size();
        std::vector<std::pair<int, int>> stack;
        std::vector<int> ans(n, 0);

        for (int i = 0; i < n; ++i) {
            while (!stack.empty() and stack.back().second < temperatures[i]) {
                auto [j, temp] = stack.back();
                stack.pop_back();
                ans[j] = i - j;
            }
            stack.emplace_back(std::make_pair(i, temperatures[i]));
        }

        return ans;
    }
};